core/http-auth-aws-crt/src/main/java/software/amazon/awssdk/http/auth/aws/crt/internal/signer/RollingSigner.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.aws.crt.internal.signer;
+
+import java.io.ByteArrayInputStream;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.crt.auth.signing.AwsSigner;
+import software.amazon.awssdk.crt.auth.signing.AwsSigningConfig;
+import software.amazon.awssdk.crt.auth.signing.AwsSigningResult;
+import software.amazon.awssdk.crt.http.HttpHeader;
+import software.amazon.awssdk.crt.http.HttpRequestBodyStream;
+import software.amazon.awssdk.http.auth.aws.crt.internal.CrtInputStream;
+
+/**
+ * A class which calculates a rolling signature of arbitrary data using HMAC-SHA256. Each time a signature is calculated, the
+ * prior calculation is incorporated, hence "rolling".
+ */
+@SdkInternalApi
+public final class RollingSigner {
+
+    private final byte[] seedSignature;
+    private final AwsSigningConfig signingConfig;
+    private byte[] previousSignature;
+
+    public RollingSigner(byte[] seedSignature, AwsSigningConfig signingConfig) {
+        this.seedSignature = seedSignature.clone();
+        this.previousSignature = seedSignature.clone();
+        this.signingConfig = signingConfig;
+    }
+
+    private static byte[] signChunk(byte[] chunkBody, byte[] previousSignature, AwsSigningConfig signingConfig) {
+        HttpRequestBodyStream crtBody = new CrtInputStream(() -> new ByteArrayInputStream(chunkBody));
```

Comment:
Do we need to close the body afterwards?

Reply:
That's a good question. Closing would need to happen somewhere downstream (such as in `AwsSigner.signChunk()`) or in `CrtInputStream` itself). It doesn't happen in `AwsSigner`, so perhaps we can add it to `CrtInputStream`.